File: src/codes/excessive-branching/py/rfc_strategy.py
from abc import ABC, abstractmethod

# Strategy インターフェース
class DiscountStrategy(ABC):
    @abstractmethod
    def calculate(self, base_price: float) -> float:
        pass

# ConcreteStrategy: 学生割引
class StudentDiscount(DiscountStrategy):
    def calculate(self, base_price: float) -> float:
        return base_price * 0.8

# ConcreteStrategy: 会員割引
class MemberDiscount(DiscountStrategy):
    def calculate(self, base_price: float) -> float:
        return base_price * 0.9

# ConcreteStrategy: VIP割引
class VIPDiscount(DiscountStrategy):
    def calculate(self, base_price: float) -> float:
        return base_price * 0.7

# ConcreteStrategy: 割引なし
class NoDiscount(DiscountStrategy):
    def calculate(self, base_price: float) -> float:
        return base_price

# Context クラス
class PriceCalculator:
    def __init__(self, discount_strategy: DiscountStrategy):
        self._discount_strategy = discount_strategy

    def calculate_price(self, base_price: float) -> float:
        return self._discount_strategy.calculate(base_price)

# 利用例
base_price = 1000

student_calculator = PriceCalculator(StudentDiscount())
print(student_calculator.calculate_price(base_price))  # 800

member_calculator = PriceCalculator(MemberDiscount())
print(member_calculator.calculate_price(base_price))  # 900

vip_calculator = PriceCalculator(VIPDiscount())
print(vip_calculator.calculate_price(base_price))  # 700

regular_calculator = PriceCalculator(NoDiscount())
print(regular_calculator.calculate_price(base_price))  # 1000

File: src/codes/excessive-branching/ts/rfc_strategy.ts
// Strategy インターフェース
interface DiscountStrategy {
  calculate(basePrice: number): number;
}

// ConcreteStrategy: 学生割引
class StudentDiscount implements DiscountStrategy {
  calculate(basePrice: number): number {
    return basePrice * 0.8;
  }
}

// ConcreteStrategy: 会員割引
class MemberDiscount implements DiscountStrategy {
  calculate(basePrice: number): number {
    return basePrice * 0.9;
  }
}

// ConcreteStrategy: VIP割引
class VIPDiscount implements DiscountStrategy {
  calculate(basePrice: number): number {
    return basePrice * 0.7;
  }
}

// ConcreteStrategy: 割引なし
class NoDiscount implements DiscountStrategy {
  calculate(basePrice: number): number {
    return basePrice;
  }
}

// Context クラス
class PriceCalculator {
  constructor(private discountStrategy: DiscountStrategy) {}

  calculatePrice(basePrice: number): number {
    return this.discountStrategy.calculate(basePrice);
  }
}

// 利用例
const basePrice = 1000;
const studentCalculator = new PriceCalculator(new StudentDiscount());
console.log(studentCalculator.calculatePrice(basePrice)); // 800

const memberCalculator = new PriceCalculator(new MemberDiscount());
console.log(memberCalculator.calculatePrice(basePrice)); // 900

const vipCalculator = new PriceCalculator(new VIPDiscount());
console.log(vipCalculator.calculatePrice(basePrice)); // 700

const regularCalculator = new PriceCalculator(new NoDiscount());
console.log(regularCalculator.calculatePrice(basePrice)); // 1000
